src/storage/restrict.rs

```diff
@@ -0,0 +1,249 @@
+{
+    type Type = (Entry<'rf, T>, Self); 
+    type Value = Self;
+    type Mask = BitSet;
+    fn open(self) -> (Self::Mask, Self::Value) {
+        (self.bitset.borrow().clone(), self)
+    }
+    unsafe fn get(value: &mut Self::Value, id: Index) -> Self::Type {
+        use std::mem;
+        let entry = Entry {
+            id: id,
+            pointer: value.data.borrow() as *const T::Storage,
+            phantom: PhantomData,
+        };
+
+        let value: &'rf mut Self::Value = mem::transmute(value);
+        (entry, value)
+    }
+}
+
+impl<'st, T, D> Storage<'st, T, D>
+    where T: Component,
+          D: Deref<Target = MaskedStorage<T>>,
+{
+    /// Builds an immutable `RestrictedStorage` out of a `Storage`. Allows restricted
+    /// access to the inner components without allowing invalidating the
+    /// bitset for iteration in `Join`.
+    pub fn restrict<'rf>(&'rf self)
+        -> RestrictedStorage<'rf, 'st, &'rf BitSet, T, &'rf T::Storage, restrict_type::Normal>
+    {
+        RestrictedStorage {
+            bitset: &self.data.mask,
+            data: &self.data.inner,
+            entities: &self.entities,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Builds a parallel `RestrictedStorage`, does not allow mutably getting other components
+    /// aside from the current iteration.
+    pub fn par_restrict<'rf>(&'rf self)
+        -> RestrictedStorage<'rf, 'st, &'rf BitSet, T, &'rf T::Storage, restrict_type::Parallel>
+    {
+        RestrictedStorage {
+            bitset: &self.data.mask,
+            data: &self.data.inner,
+            entities: &self.entities,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<'st, T, D> Storage<'st, T, D>
+    where T: Component,
+          D: DerefMut<Target = MaskedStorage<T>>,
+{
+    /// Builds a mutable `RestrictedStorage` out of a `Storage`. Allows restricted
+    /// access to the inner components without allowing invalidating the
+    /// bitset for iteration in `Join`.
+    pub fn restrict_mut<'rf>(&'rf mut self)
+        -> RestrictedStorage<'rf, 'st, &'rf BitSet, T, &'rf mut T::Storage, restrict_type::Normal>
+    {
+        let (mask, data) = self.data.open_mut();
+        RestrictedStorage {
+            bitset: mask,
+            data: data,
+            entities: &self.entities,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Builds a mutable, parallel `RestrictedStorage`, does not allow mutably getting other components
+    /// aside from the current iteration.
+    pub fn par_restrict_mut<'rf>(&'rf mut self)
+        -> RestrictedStorage<'rf, 'st, &'rf BitSet, T, &'rf mut T::Storage, restrict_type::Parallel>
+    {
+        let (mask, data) = self.data.open_mut();
+        RestrictedStorage {
+            bitset: mask,
+            data: data,
+            entities: &self.entities,
+            phantom: PhantomData,
+        }
+    }
+}
+
+/// An entry to a storage.
+pub struct Entry<'rf, T>
```

Comment:
At this point, we should evaluate which of `Send` / `Sync` this type conforms to.